File: 29Items-Projects/7Algorithms/6-Backtracking/java/src/combinations/Combinations.java
package combinations;

import java.util.*;

public class Combinations {

    private final int[] set;
    private final int k;
    private final List<List<Integer>> results = new ArrayList<>();

    public Combinations(int[] set, int k) {
        this.set = set;
        this.k = k;
    }

    public void solveAndPrint() {
        System.out.println("Set: " + Arrays.toString(set));
        System.out.println("Combination size k=" + k);
        System.out.println();

        backtrack(0, new ArrayList<>());

        System.out.printf("Found %d combination(s):%n", results.size());
        for (int i = 0; i < results.size(); i++) {
            System.out.printf("  #%d: %s%n", i + 1, results.get(i));
        }
    }

    private void backtrack(int start, List<Integer> current) {
        if (current.size() == k) {
            results.add(new ArrayList<>(current));
            return;
        }
        for (int i = start; i < set.length; i++) {
            current.add(set[i]);
            backtrack(i + 1, current);
            current.remove(current.size() - 1);
        }
    }
}

File: 29Items-Projects/7Algorithms/6-Backtracking/python/combinations.py
"""Combinations — generate all k-sized subsets from a set."""


class Combinations:
    def __init__(self, elements: list[int], k: int):
        self.elements = elements
        self.k = k
        self.results: list[list[int]] = []

    def solve_and_print(self) -> None:
        print(f"Set: {self.elements}")
        print(f"Combination size k={self.k}\n")

        self._backtrack(0, [])

        print(f"Found {len(self.results)} combination(s):")
        for i, combo in enumerate(self.results, 1):
            print(f"  #{i}: {combo}")

    def _backtrack(self, start: int, current: list[int]) -> None:
        if len(current) == self.k:
            self.results.append(current[:])
            return
        for i in range(start, len(self.elements)):
            current.append(self.elements[i])
            self._backtrack(i + 1, current)
            current.pop()
